plugins/bufferlines/bufferline.nix

```diff
@@ -125,196 +215,395 @@ in
+        Also accepts a function with a signature `function(buffer_a, buffer_b)` allowing you to compare with custom logic.
+      '';
 
-      offsets = helpers.defaultNullOpts.mkNullable (types.listOf types.attrs) null "offsets";
+      diagnostics =
+        defaultNullOpts.mkEnumFirstDefault
+          [
+            false
+            "nvim_lsp"
+            "coc"
+          ]
+          ''
+            Diagnostics provider.
+
+            Set to `false` to disable.
+          '';
+
+      diagnostics_indicator = defaultNullOpts.mkLuaFn null ''
+        Either `null` or a function that returns the diagnostics indicator.
+      '';
+
+      diagnostics_update_on_event = defaultNullOpts.mkBool true ''
+        Use nvim's diagnostic handler.
+      '';
+
+      offsets = defaultNullOpts.mkNullable (types.listOf types.attrs) null "offsets";
 
       groups = {
-        items = helpers.defaultNullOpts.mkListOf types.attrs [ ] "List of groups.";
+        items = defaultNullOpts.mkListOf types.attrs [ ] "List of groups.";
 
         options = {
-          toggleHiddenOnEnter = helpers.defaultNullOpts.mkBool true "Re-open hidden groups on bufenter.";
+          toggle_hidden_on_enter = defaultNullOpts.mkBool true ''
+            Re-open hidden groups on `BufEnter`.
+          '';
         };
       };
 
       hover = {
-        enabled = mkEnableOption "hover";
+        enabled = lib.mkEnableOption "hover";
 
-        reveal = helpers.defaultNullOpts.mkListOf types.str [ ] "reveal";
+        reveal = defaultNullOpts.mkListOf types.str [ ] "reveal";
 
-        delay = helpers.defaultNullOpts.mkInt 200 "delay";
+        delay = defaultNullOpts.mkInt 200 "delay";
       };
 
       debug = {
-        logging = helpers.defaultNullOpts.mkBool false "Whether to enable logging";
+        logging = defaultNullOpts.mkBool false "Whether to enable logging.";
       };
+    };
 
-      customFilter = helpers.defaultNullOpts.mkLuaFn null ''
-        ```
-        fun(buf: number, bufnums: number[]): boolean
-        ```
-      '';
-
-      highlights = genAttrs (attrValues highlightOptions) (
+    highlights =
+      let
+        newHighlightsOptions = map (
+          option:
+          let
+            optionSnakeCase = lib.nixvim.toSnakeCase option;
+          in
+          optionSnakeCase
+        ) (oldHighlightOptions ++ [ "truncMarker" ]);
+      in
+      lib.genAttrs newHighlightsOptions (
         name:
-        helpers.mkNullOrOption helpers.nixvimTypes.highlight ''
+        defaultNullOpts.mkHighlight { } null ''
           Highlight group definition for ${name}.
         ''
       );
+  };
+
+  extraOptions = {
+    diagnosticsUpdateInInsert = lib.mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable diagnostics update in insert mode.
+
+        Only applies when `coc` is selected in option `plugins.settings.options.diagnostic`.
+      '';
```

Comment:
I think this should be deprecated in favour of our diagnostics option: https://github.com/nix-community/nixvim/blob/dbf6f7bc997dc3a9ab1f014ea075600357226950/modules/diagnostics.nix#L10-L18

E.g:

```nix
(mkRenameOptionModule (basePluginPath ++ [ "diagnosticsUpdateInInsert" ] [
  "diagnostics"
  "update_in_insert"
])
```